src/HotChocolate/Data/src/Data/QueryableExecutable.cs

```diff
@@ -0,0 +1,47 @@
+using System.Collections;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace HotChocolate.Data
+{
+    public class QueryableExecutable<T> : IQueryableExecutable<T>
+    {
+        public QueryableExecutable(IQueryable<T> queryable)
+        {
+            Source = queryable;
+            InMemory = Source is EnumerableQuery;
+        }
+
+        public IQueryable<T> Source { get; }
+
+        object IExecutable.Source => Source;
+
+        public bool InMemory { get; }
+
+        public QueryableExecutable<T> WithSource(IQueryable<T> source)
+        {
+            return new QueryableExecutable<T>(source);
+        }
+
+        public virtual ValueTask<IList> ToListAsync(CancellationToken cancellationToken)
+        {
+            return new ValueTask<IList>(Source.ToList());
```

Comment:
This cannot be the correct implementation or?

Reply:
Even if we have a ef specific implementation there are other IQueryable drivers where we need to execute async like mongo, or ravendb and so on. So the queryable executable must check if it is in-mem and only then use sync in all other cases we must check if IAsyncEnumerable is possible or if we need to spin up a task.